In [156]:
import pandas as pd
from sqlalchemy import create_engine
import os
import numpy as np


In [157]:
csv_file = os.path.join( "Resources", "president_polls.csv")
polls_data_df = pd.read_csv(csv_file)
polls_data_df.head()

question_id  poll_id  cycle           state  pollster_id  \
0       134285    71678   2020    Pennsylvania          383   
1       134285    71678   2020    Pennsylvania          383   
2       134286    71679   2020        Michigan          383   
3       134286    71679   2020        Michigan          383   
4       134249    71662   2020  North Carolina          399   

                                   pollster sponsor_ids  \
0                     Public Policy Polling        1484   
1                     Public Policy Polling        1484   
2                     Public Policy Polling        1484   
3                     Public Policy Polling        1484   
4  Rasmussen Reports/Pulse Opinion Research         NaN   

                       sponsors                              display_name  \
0  American Bridge 21st Century                     Public Policy Polling   
1  American Bridge 21st Century                     Public Policy Polling   
2  American Bridge 21st Century                     Public Policy Polling   
3  American Bridge 21st Century                     Public Policy Polling   
4                           NaN  Rasmussen Reports/Pulse Opinion Research   

   pollster_rating_id  ...        created_at notes  \
0               263.0  ...  10/22/2020 15:33   NaN   
1               263.0  ...  10/22/2020 15:33   NaN   
2               263.0  ...  10/22/2020 15:33   NaN   
3               263.0  ...  10/22/2020 15:33   NaN   
4               277.0  ...  10/22/2020 10:27   NaN   

                                                 url    stage race_id answer  \
0  https://americanbridgepac.org/wp-content/uploa...  general    6249  Biden   
1  https://americanbridgepac.org/wp-content/uploa...  general    6249  Trump   
2  https://americanbridgepac.org/wp-content/uploa...  general    6233  Biden   
3  https://americanbridgepac.org/wp-content/uploa...  general    6233  Trump   
4  https://www.rasmussenreports.com/public_conten...  general    6238  Biden   

  candidate_id       candidate_name  candidate_party   pct  
0        13256  Joseph R. Biden Jr.              DEM  51.0  
1        13254         Donald Trump              REP  46.0  
2        13256  Joseph R. Biden Jr.              DEM  50.0  
3        13254         Donald Trump              REP  43.0  
4        13256  Joseph R. Biden Jr.              DEM  48.0  

[5 rows x 38 columns]

In [158]:
new_polls_data_df = polls_data_df[[ 'state', 'pollster_id', 'pollster', 'fte_grade', 'sample_size', 'population', 'methodology', 'end_date', 'nationwide_batch', 'answer', 'candidate_party', 'pct' ]].copy()
new_polls_data_df.head()

state  pollster_id                                  pollster  \
0    Pennsylvania          383                     Public Policy Polling   
1    Pennsylvania          383                     Public Policy Polling   
2        Michigan          383                     Public Policy Polling   
3        Michigan          383                     Public Policy Polling   
4  North Carolina          399  Rasmussen Reports/Pulse Opinion Research   

  fte_grade  sample_size population methodology    end_date  nationwide_batch  \
0         B        890.0          v    IVR/Text  10/22/2020             False   
1         B        890.0          v    IVR/Text  10/22/2020             False   
2         B        804.0          v    IVR/Text  10/22/2020             False   
3         B        804.0          v    IVR/Text  10/22/2020             False   
4        C+        800.0         lv  IVR/Online  10/21/2020             False   

  answer candidate_party   pct  
0  Biden             DEM  51.0  
1  Trump             REP  46.0  
2  Biden             DEM  50.0  
3  Trump             REP  43.0  
4  Biden             DEM  48.0

In [159]:
# new_polls_data_df['row_id'] = np.arange(len(new_polls_data_df))
# new_polls_data_df.head()


# Which colums have null data?

In [160]:
new_polls_data_df.isnull().any()

state                True
pollster_id         False
pollster            False
fte_grade            True
sample_size          True
population          False
methodology          True
end_date            False
nationwide_batch    False
answer              False
candidate_party     False
pct                 False
dtype: bool

**Columns with null values: state, fte_grade, sample_size, and methodology**

In [161]:
new_polls_data_df.count()

state                8184
pollster_id         12812
pollster            12812
fte_grade           11865
sample_size         12810
population          12812
methodology         12783
end_date            12812
nationwide_batch    12812
answer              12812
candidate_party     12812
pct                 12812
dtype: int64

In [162]:
new_polls_data_df.isnull().any(axis=1)

0        False
1        False
2        False
3        False
4        False
         ...  
12807    False
12808    False
12809    False
12810     True
12811     True
Length: 12812, dtype: bool

In [163]:
new_polls_data_df[new_polls_data_df.isnull().any(axis=1)]

state  pollster_id                                  pollster fte_grade  \
8       NaN         1193                              SurveyMonkey        D-   
9       NaN         1193                              SurveyMonkey        D-   
10      NaN         1193                              SurveyMonkey        D-   
11      NaN         1193                              SurveyMonkey        D-   
12      NaN          228                                  IBD/TIPP       A/B   
...     ...          ...                                       ...       ...   
12799   NaN         1416               Harris Insights & Analytics         C   
12800   NaN         1416               Harris Insights & Analytics         C   
12801   NaN         1416               Harris Insights & Analytics         C   
12810   NaN          399  Rasmussen Reports/Pulse Opinion Research        C+   
12811   NaN          399  Rasmussen Reports/Pulse Opinion Research        C+   

       sample_size population        methodology    end_date  \
8          13943.0         rv             Online  10/21/2020   
9          13943.0         rv             Online  10/21/2020   
10         13943.0         lv             Online  10/21/2020   
11         13943.0         lv             Online  10/21/2020   
12           965.0         lv  Live Phone/Online  10/21/2020   
...            ...        ...                ...         ...   
12799       1001.0         rv             Online  12/17/2018   
12800       1001.0         rv             Online  12/17/2018   
12801       1001.0         rv             Online  12/17/2018   
12810       1000.0         lv         IVR/Online  11/13/2018   
12811       1000.0         lv         IVR/Online  11/13/2018   

       nationwide_batch    answer candidate_party   pct  
8                 False     Biden             DEM  52.0  
9                 False     Trump             REP  45.0  
10                False     Biden             DEM  52.0  
11                False     Trump             REP  46.0  
12                False     Biden             DEM  49.6  
...                 ...       ...             ...   ...  
12799             False     Trump             REP  37.0  
12800             False  O'Rourke             DEM  30.0  
12801             False     Trump             REP  37.0  
12810             False     Obama             DEM  50.0  
12811             False     Trump             REP  43.0  

[5215 rows x 12 columns]

**There are 5,215 rows out of 12,812 rows with null data. Since we are interested in state polling data in relations to our financial data by state, we will drop all rows in polls.csv that contain null values in state column.**

In [164]:
new_polls_data_df = new_polls_data_df.dropna(axis=0 , how='any', subset = ['state'])

In [165]:
new_polls_data_df.isnull().any()

state               False
pollster_id         False
pollster            False
fte_grade            True
sample_size         False
population          False
methodology          True
end_date            False
nationwide_batch    False
answer              False
candidate_party     False
pct                 False
dtype: bool

In [166]:
new_polls_data_df.count()

state               8184
pollster_id         8184
pollster            8184
fte_grade           7620
sample_size         8184
population          8184
methodology         8159
end_date            8184
nationwide_batch    8184
answer              8184
candidate_party     8184
pct                 8184
dtype: int64

# Add a new column called committee_id for Trump and Biden

In [167]:
new_polls_data_df['committee_id']=''
new_polls_data_df.head()

state  pollster_id                                  pollster  \
0    Pennsylvania          383                     Public Policy Polling   
1    Pennsylvania          383                     Public Policy Polling   
2        Michigan          383                     Public Policy Polling   
3        Michigan          383                     Public Policy Polling   
4  North Carolina          399  Rasmussen Reports/Pulse Opinion Research   

  fte_grade  sample_size population methodology    end_date  nationwide_batch  \
0         B        890.0          v    IVR/Text  10/22/2020             False   
1         B        890.0          v    IVR/Text  10/22/2020             False   
2         B        804.0          v    IVR/Text  10/22/2020             False   
3         B        804.0          v    IVR/Text  10/22/2020             False   
4        C+        800.0         lv  IVR/Online  10/21/2020             False   

  answer candidate_party   pct committee_id  
0  Biden             DEM  51.0               
1  Trump             REP  46.0               
2  Biden             DEM  50.0               
3  Trump             REP  43.0               
4  Biden             DEM  48.0

In [168]:
new_polls_data_df['answer'].value_counts()

Trump           3796
Biden           2965
Jorgensen        282
Sanders          260
Warren           210
Hawkins          175
Buttigieg        138
Harris            79
Bloomberg         58
Klobuchar         43
O'Rourke          28
West              25
Pence             22
Booker            15
Blankenship       13
Schultz           12
Haley             10
De La Fuente       8
Castro             7
Yang               6
Steyer             5
Pierce             5
Simmons            3
Amash              2
La Riva            2
Gillibrand         2
Winfrey            2
Brown              2
Gabbard            2
Obama              2
Pelosi             1
Kennedy            1
de Blasio          1
Bullock            1
Williamson         1
Name: answer, dtype: int64

In [169]:
new_polls_data_df.loc[new_polls_data_df['answer'] == 'Trump', 'committee_id'] = "C00580100"
new_polls_data_df.loc[new_polls_data_df['answer'] == 'Biden', 'committee_id'] = "C00703975"
new_polls_data_df.head()

state  pollster_id                                  pollster  \
0    Pennsylvania          383                     Public Policy Polling   
1    Pennsylvania          383                     Public Policy Polling   
2        Michigan          383                     Public Policy Polling   
3        Michigan          383                     Public Policy Polling   
4  North Carolina          399  Rasmussen Reports/Pulse Opinion Research   

  fte_grade  sample_size population methodology    end_date  nationwide_batch  \
0         B        890.0          v    IVR/Text  10/22/2020             False   
1         B        890.0          v    IVR/Text  10/22/2020             False   
2         B        804.0          v    IVR/Text  10/22/2020             False   
3         B        804.0          v    IVR/Text  10/22/2020             False   
4        C+        800.0         lv  IVR/Online  10/21/2020             False   

  answer candidate_party   pct committee_id  
0  Biden             DEM  51.0    C00703975  
1  Trump             REP  46.0    C00580100  
2  Biden             DEM  50.0    C00703975  
3  Trump             REP  43.0    C00580100  
4  Biden             DEM  48.0    C00703975

# Evaluate what type of data types are there?

In [170]:
new_polls_data_df.dtypes

state                object
pollster_id           int64
pollster             object
fte_grade            object
sample_size         float64
population           object
methodology          object
end_date             object
nationwide_batch       bool
answer               object
candidate_party      object
pct                 float64
committee_id         object
dtype: object

In [171]:
output_file = os.path.join("Output", "polls_clean_df.csv")

new_polls_data_df.to_csv(output_file, index=False, header=True)

# Connect to local database

In [172]:
pg_user = 'postgres'
pg_password = 'Rouzbeh237*tx'
db_name = 'POTUS_db'

connection_string = f"{pg_user}:{pg_password}@localhost:5432/{db_name}"
engine = create_engine(f'postgresql://{connection_string}')

In [173]:
engine.table_names()

['States', 'Polls', 'Candidates', 'Candidate_Spending']

In [174]:
new_polls_data_df.to_sql(name="Polls", con=engine, if_exists='append', index=False)

IntegrityError: (psycopg2.errors.ForeignKeyViolation) insert or update on table "Polls" violates foreign key constraint "fk_Polls_committee_id"
DETAIL:  Key (committee_id)=() is not present in table "Candidates".

[SQL: INSERT INTO "Polls" (state, pollster_id, pollster, fte_grade, sample_size, population, methodology, end_date, nationwide_batch, answer, candidate_party, pct, committee_id) VALUES (%(state)s, %(pollster_id)s, %(pollster)s, %(fte_grade)s, %(sample_size)s, %(population)s, %(methodology)s, %(end_date)s, %(nationwide_batch)s, %(answer)s, %(candidate_party)s, %(pct)s, %(committee_id)s)]
[parameters: ({'state': 'Pennsylvania', 'pollster_id': 383, 'pollster': 'Public Policy Polling', 'fte_grade': 'B', 'sample_size': 890.0, 'population': 'v', 'methodology': 'IVR/Text', 'end_date': '10/22/2020', 'nationwide_batch': False, 'answer': 'Biden', 'candidate_party': 'DEM', 'pct': 51.0, 'committee_id': 'C00703975'}, {'state': 'Pennsylvania', 'pollster_id': 383, 'pollster': 'Public Policy Polling', 'fte_grade': 'B', 'sample_size': 890.0, 'population': 'v', 'methodology': 'IVR/Text', 'end_date': '10/22/2020', 'nationwide_batch': False, 'answer': 'Trump', 'candidate_party': 'REP', 'pct': 46.0, 'committee_id': 'C00580100'}, {'state': 'Michigan', 'pollster_id': 383, 'pollster': 'Public Policy Polling', 'fte_grade': 'B', 'sample_size': 804.0, 'population': 'v', 'methodology': 'IVR/Text', 'end_date': '10/22/2020', 'nationwide_batch': False, 'answer': 'Biden', 'candidate_party': 'DEM', 'pct': 50.0, 'committee_id': 'C00703975'}, {'state': 'Michigan', 'pollster_id': 383, 'pollster': 'Public Policy Polling', 'fte_grade': 'B', 'sample_size': 804.0, 'population': 'v', 'methodology': 'IVR/Text', 'end_date': '10/22/2020', 'nationwide_batch': False, 'answer': 'Trump', 'candidate_party': 'REP', 'pct': 43.0, 'committee_id': 'C00580100'}, {'state': 'North Carolina', 'pollster_id': 399, 'pollster': 'Rasmussen Reports/Pulse Opinion Research', 'fte_grade': 'C+', 'sample_size': 800.0, 'population': 'lv', 'methodology': 'IVR/Online', 'end_date': '10/21/2020', 'nationwide_batch': False, 'answer': 'Biden', 'candidate_party': 'DEM', 'pct': 48.0, 'committee_id': 'C00703975'}, {'state': 'North Carolina', 'pollster_id': 399, 'pollster': 'Rasmussen Reports/Pulse Opinion Research', 'fte_grade': 'C+', 'sample_size': 800.0, 'population': 'lv', 'methodology': 'IVR/Online', 'end_date': '10/21/2020', 'nationwide_batch': False, 'answer': 'Trump', 'candidate_party': 'REP', 'pct': 49.0, 'committee_id': 'C00580100'}, {'state': 'Iowa', 'pollster_id': 1102, 'pollster': 'Emerson College', 'fte_grade': 'A-', 'sample_size': 435.0, 'population': 'lv', 'methodology': 'IVR/Online/Text', 'end_date': '10/21/2020', 'nationwide_batch': False, 'answer': 'Biden', 'candidate_party': 'DEM', 'pct': 48.1, 'committee_id': 'C00703975'}, {'state': 'Iowa', 'pollster_id': 1102, 'pollster': 'Emerson College', 'fte_grade': 'A-', 'sample_size': 435.0, 'population': 'lv', 'methodology': 'IVR/Online/Text', 'end_date': '10/21/2020', 'nationwide_batch': False, 'answer': 'Trump', 'candidate_party': 'REP', 'pct': 48.0, 'committee_id': 'C00580100'}  ... displaying 10 of 8184 total bound parameter sets ...  {'state': 'Ohio', 'pollster_id': 383, 'pollster': 'Public Policy Polling', 'fte_grade': 'B', 'sample_size': 648.0, 'population': 'v', 'methodology': 'Automated Phone', 'end_date': '11/28/2018', 'nationwide_batch': False, 'answer': 'Warren', 'candidate_party': 'DEM', 'pct': 43.0, 'committee_id': ''}, {'state': 'Ohio', 'pollster_id': 383, 'pollster': 'Public Policy Polling', 'fte_grade': 'B', 'sample_size': 648.0, 'population': 'v', 'methodology': 'Automated Phone', 'end_date': '11/28/2018', 'nationwide_batch': False, 'answer': 'Trump', 'candidate_party': 'REP', 'pct': 49.0, 'committee_id': 'C00580100'})]
(Background on this error at: http://sqlalche.me/e/gkpj)

In [155]:
pd.read_sql_query('select * from "Polls"', con=engine).head()

Empty DataFrame
Columns: [state, pollster_id, pollster, fte_grade, sample_size, population, methodology, end_date, nationwide_batch, answer, candidate_party, pct, committee_id, id]
Index: []